### DISCUSSION TITLE
Can you inherit from BaseModel and Exception?

### DISCUSSION THREAD

**Discussion by ch1nq** [pydantic/pydantic#5770] [Question] [answered]
I am trying to implement a custom exception type with custom fields that can be overwritten and (de)serialized. I have tried to use inherit from BaseModel:
```python
class ParentException(Exception, BaseModel):
    custom_field: set[str] = pydantic.Field(default=set())
```
However, inheriting from `Exception` and `BaseError` get a layout conflict 
```
TypeError: multiple bases have instance lay-out conflict
```

I have also tried using a dataclass instead, however, i have issues with overwriting fields in subclasses:

```python
@pydantic.dataclasses.dataclass
class ParentException(Exception):
    custom_field: str = pydantic.Field(default="parent")

class ChildException(ParentException):
    custom_field: str = "child"

assert ChildException().custom_field == "child" # fails since custom_field is "parent"
```

Is there a way to acheive this?

**Comment by eigenein** [comment]
👋 I was hoping for v2 to resolve it, but the lay-out conflict is still present in `2.3.0`. Any workarounds for v2 maybe?

**Comment by harunyasar** [comment, marked as answer]
```python
from pydantic import Field
from pydantic.dataclasses import dataclass


@dataclass
class ParentException(Exception):
    custom_field: str = Field(default="parent")

@dataclass
class ChildException(ParentException):
    custom_field: str = "child"

assert ChildException().custom_field == "child"
```

This should work.

**Reply by eigenein** [reply]
Unfortunately, this wouldn't work with the «normal» inheritance from `BaseModel`. I'd love to see it work like this:

```python
class ParentException(BaseModel, Exception):
    …
```

**Comment by eigenein** [comment]
@samuelcolvin 👋 I'm wondering, is there a good reason to have `__slots__` in `BaseModel` here? Because having it is the root cause of the `TypeError` in question:

https://github.com/pydantic/pydantic/blob/b7d92817ac2223237b5d341227de3eff7a84b350/pydantic/main.py#L148

There is a handful of the other `self.__pydantic_*__` attributes, which are not included in `__slots__`, which in turn makes me wonder why the included ones are so special (besides `__dict__`, of course)

If we could just remove the `__slots__` from `BaseModel`, then it could support that multiple inheritance with slots-using classes. Do you see any downsides to it?

It was [already suggested in the past for 1.x](https://github.com/pydantic/pydantic/issues/1875#issuecomment-752849476), but I believe it wasn't properly considered

It seems that these `__slots__` made their beginning [here 6 years ago](https://github.com/pydantic/pydantic/commit/6923e5e677362adbe3a1ce93fa9f776080b8d904), but I don't see an explanation of why it was added. Do you maybe still recall it? Does the original reason still hold?

**Reply by eigenein** [reply]
I'm beginning to see the fundamental problem: Pydantic isolates `__pydantic_fields_set__`, `__pydantic_extra__`, and `__pydantic_private__` from `BaseModel` instance's `__dict__` by putting them into the `__slots__`, and treats the `__dict__` entries as the model fields. The rest of the dunder-attributes are class variables and hence, no problem for the instance's `__dict__`

Changing that is not a simple matter of [dropping the `__slots__`](https://github.com/pydantic/pydantic/pull/8040), as it would require filtering `__dict__` entries whenever field values are needed, and I'm not sure now whether it's justified and a safe thing to do (and it likely would have performance implications)

**Reply by lord-haffi** [reply]
@eigenein Stumbled across this issue today. Did you make any progress or had any more insights regarding this issue?

Your explanation makes perfectly sense. I tried out the dataclass workaround but unfortunately, pydantic doesn't seem to support overridden constructors on it. I.e. I wasn't able to set `Exception.args`. I also tried to create a new BaseModel using [this hack](https://github.com/pydantic/pydantic/issues/1875#issuecomment-752855246) but it doesn't work with pydantic v2 (actually it surprises me that it worked in v1).

I think for now I will go with an approach similar to [this one](https://github.com/pydantic/pydantic/issues/1875#issuecomment-783459329). But it would be very neat if we could make this work.

The question which I'm asking me is: Why did they design the `__dict__` attribute as a dictionary containing all attributes of the model in the first place? Does it come with any performance boost if you're using `__dict__` instead of an extra e.g.  `__model_dict__` when iterating? Or was it just some kind of "eye candy" to simply throw `dir(my_model)` on it?

**Comment by lord-haffi** [comment]
After diggin into this problem for several hours I finally came up with some dirty stuff to make it work good enough for me. And it only involves an additional metaclass you have to define on each of your error models.

## What the solution can
- Errors can be instantiated just like pydantic `BaseModel`s
- The Exception `args` and `notes` can be defined through dunder instance methods
- Instances of your error model can be raised like a normal `Exception`
- Instances of your error model can be serialized like a normal `BaseModel`
- The error model can also correctly deserialize i.e. it can do a full roundtrip
- Errors can be broadly catched by superclasses

## What it cannot
It might won't work for every edge case. Also, `mypy` surely have some complains here, but I didn't care at this point. For me it did the job.

## How it looks like in action

```python
class SomeSuperClassError(SerializableErrorBase, metaclass=SerializableErrorMeta): ...

class SomeSpecificError(SomeSuperClassError, metaclass=SerializableErrorMeta):
    type: Literal["SomeSpecificError"] = "SomeSpecificError"
    x: int
    y: str

    def __exception_args__(self) -> Iterable[Any]:
        return f"There was some error with x={self.x}",

    def __exception_notes__(self) -> Iterable[str]:
        return "Note 1", "Note 2"

my_error = SomeSpecificError(x=1, y="What did it cost...")
try:
    raise my_error
except SomeSuperClassError as catched_error:
    pass
serialized_error = my_error.model_dump_json(indent=2)
deserialized_error = my_error.model_validate_json(serialized_error)
assert my_error == deserialized_error

print(deserialized_error.y)
print("Everything.")
```

## The solution
I know, it looks... stressfull. But it does the job. Enjoy :)
(Note: There might be some imports missing)

<details>
<summary>Click here to see the big chunk of code</summary>

```python
import operator
from typing import Any, Iterable
from pydantic import BaseModel, computed_field, model_validator
from pydantic._internal import _utils
from pydantic._internal._model_construction import ModelMetaclass


class MyModelMetaclass(ModelMetaclass):
    def __instancecheck__(self, instance: Any) -> bool:
        return (
            super().__instancecheck__(instance)
            or hasattr(instance, "_pydantic_model")
            and isinstance(instance._pydantic_model, self)
        )


# https://github.com/pydantic/pydantic/issues/1875#issuecomment-2707710500
class SerializableErrorMeta(ModelMetaclass):
    def __new__(mcs, name: str, bases: tuple[type, ...], namespace: dict[str, Any]):
        is_base_class = name == "SerializableErrorBase"
        if is_base_class:
            pydantic_bases = (BaseModel,)
            bases = (Exception,)
        else:
            # If a base class is an ExceptionProxy we need to subclass from the pydantic class instead
            pydantic_bases = tuple(
                (
                    base.__pydantic_model_class__
                    if issubclass(base, Exception) and hasattr(base, "__pydantic_model_class__")
                    else base
                )
                for base in bases
            )
        pydantic_cls = MyModelMetaclass.__new__(MyModelMetaclass, name, pydantic_bases, namespace)

        if is_base_class:

            __slots__ = ("_pydantic_model",)

            def __new__(cls, *, __from_pydantic__: "SerializableErrorBase | None" = None, **data):
                if __from_pydantic__ is not None:
                    if hasattr(__from_pydantic__, "_error_proxy"):
                        return __from_pydantic__._error_proxy
                    instance = Exception.__new__(cls)
                    instance._pydantic_model = __from_pydantic__
                else:
                    instance = Exception.__new__(cls)
                    instance._pydantic_model = instance.__pydantic_model_class__(**data)
                instance._pydantic_model._error_proxy = instance
                instance.__dict__ = instance._pydantic_model.__dict__
                Exception.__init__(instance, *instance._pydantic_model.__exception_args__())
                for note in instance._pydantic_model.__exception_notes__():
                    instance.add_note(note)
                return instance

            def __init__(self, *args, **kwargs):
                pass

            def __getattr__(self, item):
                try:
                    return getattr(self._pydantic_model, item)
                except AttributeError:
                    return Exception.__getattribute__(self, item)

            def __repr__(self):
                return self._pydantic_model.__repr__()

            def __eq__(self, other):
                if not hasattr(other, "_pydantic_model"):
                    return False
                getter = operator.itemgetter(
                    *(type(self._pydantic_model).__pydantic_fields__.keys() or (lambda _: _utils._SENTINEL,))
                )
                return getter(self.__dict__) == getter(other.__dict__)

            ErrorProxy = type.__new__(
                mcs,
                name + "Proxy",
                bases,
                {
                    "__slots__": __slots__,
                    "__pydantic_model_class__": pydantic_cls,
                    "__new__": __new__,
                    "__init__": __init__,
                    "__getattr__": __getattr__,
                    "__repr__": __repr__,
                    "__eq__": __eq__,
                    # "__module__": pydantic_cls.__module__,
                    # "__pydantic_init_subclass__": classmethod(lambda *_, **__: None),
                },
            )

        else:
            ErrorProxy = type.__new__(
                mcs,
                name + "Proxy",
                bases,
                {"__pydantic_model_class__": pydantic_cls, "__module__": pydantic_cls.__module__},
            )

        pydantic_cls.__error_proxy_class__ = ErrorProxy

        return ErrorProxy

    def __getattr__(cls, item):
        try:
            return getattr(cls.__pydantic_model_class__, item)
        except AttributeError:
            return super().__getattribute__(item)

    def __instancecheck__(cls, instance: Any) -> bool:
        return hasattr(instance, "_pydantic_model") and isinstance(
            instance._pydantic_model, cls.__pydantic_model_class__
        )


class SerializableErrorBase(BaseModel, Exception, metaclass=SerializableErrorMeta):
    __error_proxy_class__: ClassVar[type]
    _error_proxy: Exception

    def __exception_args__(self) -> Iterable[Any]:
        return ()

    def __exception_notes__(self) -> Iterable[str]:
        return ()

    @computed_field
    @property
    def detail(self) -> str:
        return self._error_proxy.__str__()

    @model_validator(mode="after")
    def deserialize_to_proxy(self) -> "SerializableErrorBase":
        return self.__error_proxy_class__(__from_pydantic__=self)
```
</details>

### Why the dataclass approach didn't work for me
The two most important issues I had:
1. Pydantic dataclasses don't support custom `__init__` functions
2. Inheritance with dataclasses doesn't work (at least the way I want, didn't look at it in detail)

**Reply by eigenein** [reply]
Hi @lord-haffi and sorry for the delayed response.

Looks like you're way ahead of me :) I haven't found any reasonable workaround and went ahead with kind of [«derived» exception classes](https://github.com/kpn/combadge/blob/621f3133fe4273283c6ae3d37c2caf1eaba7c968/combadge/core/response.py#L120-L179). TL;DR: `__init_subclass__()` defines a separate exception class which is then attached as the `Error` class attribute of the model

**Reply by lord-haffi** [reply]
Hi @eigenein, no stress, didn't expect you to answer that fast tbh ^^ I kind of took it as a challenge to create an object which can act both as `Exception` and as `BaseModel`. Also wanted to learn more about this whole stuff `__slots__`, `mro`, ...
Your workaround definetely looks less complicated and may be a better solution in big projects or for some special edge cases. Thanks for sharing :)
